reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,85 +83,116 @@ object StreamSubscriber {
+          case Receiving(_)         => UpstreamCompletion -> q.enqueue1(None.asRight)
+          case _                    => UpstreamCompletion -> F.unit
         }
         case OnError(e) => {
-          case WaitingOnUpstream(_, r) => UpstreamError(e) -> r.complete(e.asLeft)
-          case _                       => UpstreamError(e) -> F.unit
+          case WaitingOnUpstream(_) =>
+            UpstreamError(e) -> (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight))
+          case Receiving(_) => UpstreamCompletion -> q.enqueue1(None.asRight)
+          case _            => UpstreamError(e) -> F.unit
         }
         case OnFinalize => {
-          case WaitingOnUpstream(sub, r) =>
-            DownstreamCancellation -> (F.delay(sub.cancel) >> r.complete(None.asRight))
+          case WaitingOnUpstream(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
           case Idle(sub) => DownstreamCancellation -> F.delay(sub.cancel)
           case o         => o -> F.unit
         }
-        case OnDequeue(r) => {
-          case Uninitialized          => RequestBeforeSubscription(r) -> F.unit
-          case Idle(sub)              => WaitingOnUpstream(sub, r) -> F.delay(sub.request(1))
-          case err @ UpstreamError(e) => err -> r.complete(e.asLeft)
-          case UpstreamCompletion     => UpstreamCompletion -> r.complete(None.asRight)
-          case o                      => o -> r.complete((new Error(s"received request in invalid state [$o]")).asLeft)
+        case OnDequeue => {
+          case Uninitialized =>
+            RequestBeforeSubscription -> F.unit
+          case Idle(sub) =>
+            WaitingOnUpstream(sub) -> F.delay(sub.request(batchSize))
+          case NextBatch(sub) =>
+            Receiving(sub) -> F.delay(sub.request(batchSize))
+          case Receiving(sub) =>
+            Receiving(sub) -> F.unit
+          case err @ UpstreamError(e) =>
+            err -> (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight))
+          case UpstreamCompletion =>
+            UpstreamCompletion -> q.enqueue1(None.asRight)
+          case o =>
+            o -> (q.enqueue1((new Error(s"received request in invalid state [$o]")).asLeft) >> q
+              .enqueue1(None.asRight))
         }
       }
 
-    Ref.of[F, State](Uninitialized).map { ref =>
-      new FSM[F, A] {
-        def nextState(in: Input): F[Unit] = ref.modify(step(in)).flatten
-        def onSubscribe(s: Subscription): F[Unit] = nextState(OnSubscribe(s))
-        def onNext(a: A): F[Unit] = nextState(OnNext(a))
-        def onError(t: Throwable): F[Unit] = nextState(OnError(t))
-        def onComplete: F[Unit] = nextState(OnComplete)
-        def onFinalize: F[Unit] = nextState(OnFinalize)
-        def dequeue1: F[Either[Throwable, Option[A]]] =
-          Deferred[F, Out].flatMap(p => ref.modify(step(OnDequeue(p))).flatten >> p.get)
-      }
+    (Ref.of[F, State](Uninitialized), Ref.of[F, Long](0)).mapN {
```

Comment:
I'll see if I can get rid of this counter. Maybe `offer1` to the queue and if it's `false` I can request the next batch?

Reply:
I've been thinking about this, and I haven't decided one way or the other. If we use the `counter`, we have to get and update it in various places, which decentralizes the logic that the counter represents.

And so, the `offer1` idea is nicer because we then we don't need extra explicit state; the state is instead held as part of the queue state (i.e. how many elements it currently contains).

Reply:
I think dropping the counter altogether in favor of requesting the size of the `Chunk` produced would be preferable, something akin to `q.dequeueChunk1(size).flatMap(c => demandMore(c.size).as(c))`.

Reply:
> q.dequeueChunk1(size).flatMap(c => demandMore(c.size).as(c))

:+1: Totally agree, this works nicely at current mongo-ds implementation